rpc/lib/client/ws_client.go

```diff
@@ -1,160 +1,390 @@
+	c.ErrorsCh = make(chan error)
+
+	c.send = make(chan types.RPCRequest)
+	// 1 additional error may come from the read/write
+	// goroutine depending on which failed first.
+	c.reconnectAfter = make(chan error, 1)
+	// capacity for 1 request. a user won't be able to send more because the send
+	// channel is unbuffered.
+	c.backlog = make(chan types.RPCRequest, 1)
+
+	c.startReadWriteRoutines()
+	go c.reconnectRoutine()
+
 	return nil
 }
 
-// OnReset implements cmn.BaseService interface
-func (wsc *WSClient) OnReset() error {
-	return nil
+// OnStop implements cmn.Service.
+func (c *WSClient) OnStop() {}
+
+// Stop overrides cmn.Service#Stop. There is no other way to wait until Quit
+// channel is closed.
+func (c *WSClient) Stop() bool {
+	success := c.BaseService.Stop()
+	// only close user-facing channels when we can't write to them
+	c.wg.Wait()
+	close(c.ResultsCh)
+	close(c.ErrorsCh)
+	return success
+}
+
+// IsReconnecting returns true if the client is reconnecting right now.
+func (c *WSClient) IsReconnecting() bool {
+	return c.reconnecting
+}
+
+// IsActive returns true if the client is running and not reconnecting.
+func (c *WSClient) IsActive() bool {
+	return c.IsRunning() && !c.IsReconnecting()
+}
+
+// Send asynchronously sends the given RPCRequest to the server. Results will
+// be available on ResultsCh, errors, if any, on ErrorsCh.
+func (c *WSClient) Send(ctx context.Context, request types.RPCRequest) error {
+	select {
+	case c.send <- request:
+		c.Logger.Info("sent a request", "req", request)
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Call asynchronously calls a given method by sending an RPCRequest to the
+// server. Results will be available on ResultsCh, errors, if any, on ErrorsCh.
+func (c *WSClient) Call(ctx context.Context, method string, params map[string]interface{}) error {
+	request, err := types.MapToRequest("", method, params)
+	if err != nil {
+		return err
+	}
+	return c.Send(ctx, request)
+}
+
+// CallWithArrayParams asynchronously calls a given method by sending an
+// RPCRequest to the server. Results will be available on ResultsCh, errors, if
+// any, on ErrorsCh.
+func (c *WSClient) CallWithArrayParams(ctx context.Context, method string, params []interface{}) error {
+	request, err := types.ArrayToRequest("", method, params)
+	if err != nil {
+		return err
+	}
+	return c.Send(ctx, request)
 }
 
-func (wsc *WSClient) dial() error {
+///////////////////////////////////////////////////////////////////////////////
+// Private methods
 
-	// Dial
+func (c *WSClient) dial() error {
 	dialer := &websocket.Dialer{
-		NetDial: wsc.Dialer,
+		NetDial: c.Dialer,
 		Proxy:   http.ProxyFromEnvironment,
 	}
 	rHeader := http.Header{}
-	con, _, err := dialer.Dial("ws://"+wsc.Address+wsc.Endpoint, rHeader)
+	conn, _, err := dialer.Dial("ws://"+c.Address+c.Endpoint, rHeader)
 	if err != nil {
 		return err
 	}
-	// Set the ping/pong handlers
-	con.SetPingHandler(func(m string) error {
```

Comment:
The default handler is enough, I think. By using `send` channel (unbuffered) we implement a simple backoff for any user. This should guaranty us write buffer won't be filled.